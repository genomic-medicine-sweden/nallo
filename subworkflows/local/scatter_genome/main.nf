include { BEDTOOLS_MAKEWINDOWS     } from '../../../modules/nf-core/bedtools/makewindows/main'
include { BEDTOOLS_MERGE           } from '../../../modules/nf-core/bedtools/merge/main'
include { BEDTOOLS_SORT            } from '../../../modules/nf-core/bedtools/sort/main'
include { BUILD_INTERVALS          } from '../../../modules/local/build_intervals/main'
include { SPLIT_BED_CHUNKS         } from '../../../modules/local/split_bed_chunks/main'

workflow SCATTER_GENOME {

    take:
    ch_fai             // channel: [optional] [ val(meta), path(fai) ]
    ch_input_bed       // channel: [optional] [ val(meta), path(bed) ]
    make_bed_from_fai  // bool
    make_bed_intervals // bool
    split_n            // integer: split bed into n regions

    main:
    ch_versions = Channel.empty()
    ch_bed = Channel.empty()
    ch_bed_intervals = Channel.empty()

    //
    // If no BED-file is provided then build intervals from reference
    //
    if( make_bed_from_fai ) {


        BUILD_INTERVALS ( ch_fai.map { id, fai -> [ [ 'id': id ], fai ] } )
        ch_versions = ch_versions.mix(BUILD_INTERVALS.out.versions)

        BUILD_INTERVALS.out.bed
            .set{ ch_bed }
    } else {
        ch_input_bed
            .set{ ch_bed }
    }

    //
    // Merge overlapping and then split BED regions for SNV calling
    //
    if( make_bed_intervals ) {

        if( split_n < 1 ) { error "Can't split bed file into less than one file" }

        // Sort and merge overlapping regions
        BEDTOOLS_SORT ( ch_bed, [] )
        ch_versions = ch_versions.mix(BEDTOOLS_SORT.out.versions)

        BEDTOOLS_MERGE ( BEDTOOLS_SORT.out.sorted )
        ch_versions = ch_versions.mix(BEDTOOLS_MERGE.out.versions)

        BEDTOOLS_MAKEWINDOWS ( BEDTOOLS_MERGE.out.bed )
        ch_versions = ch_versions.mix(BEDTOOLS_MAKEWINDOWS.out.versions)

        SPLIT_BED_CHUNKS( BEDTOOLS_MAKEWINDOWS.out.bed, split_n )
        ch_versions = ch_versions.mix(SPLIT_BED_CHUNKS.out.versions)

        // Create a channel with the bed file and the total number of intervals (for groupKey)
        SPLIT_BED_CHUNKS.out.split_beds
            .collect()
            .map{ it -> [ it, it.size() ] }
            .transpose()
            .set { ch_bed_intervals }
    }

    emit:
    bed           = ch_bed                             // channel: [ val(meta), path(bed) ]
    bed_intervals = ch_bed_intervals                   // channel: [ path(bed), val(num_intervals) ]
    versions      = ch_versions                        // channel: [ versions.yml ]
}

