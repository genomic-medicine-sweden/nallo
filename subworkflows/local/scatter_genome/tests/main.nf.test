nextflow_workflow {

    name "Test Workflow SCATTER_GENOME"
    script "../main.nf"
    workflow "SCATTER_GENOME"
    config "./nextflow.config"

    setup {
        run("SAMTOOLS_FAIDX") {
            script "../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = [[],[]]
                """
            }
        }
    }

    test("fai, [], true, false, []") {

        when {
            workflow {
                """
                input[0] = SAMTOOLS_FAIDX.out.fai
                input[1] = [[],[]]
                input[2] = true
                input[3] = false
                input[4] = []
                """
            }
        }

        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    path(workflow.out.bed.get(0).get(1)).readLines(),
                    workflow.out.bed_intervals.findAll { it[0] instanceof String }.collect { path(it[0]).readLines() },
                ).match() }
            )
        }

    }

    test("fai, [], true, true, 3") {

        when {
            workflow {
                """
                input[0] = SAMTOOLS_FAIDX.out.fai
                input[1] = [[],[]]
                input[2] = true
                input[3] = true
                input[4] = 3
                """
            }
        }

        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    path(workflow.out.bed.get(0).get(1)).readLines(),
                    workflow.out.bed_intervals.findAll { it[0] instanceof String }.collect { path(it[0]).readLines() },
                ).match() }
            )
        }

    }

    test("fai, [], bed, true, 3") {

        when {
            workflow {
                """
                input[0] = SAMTOOLS_FAIDX.out.fai
                input[1] = [[],[]]
                input[2] = Channel.of('chr22\t0\t1000')
                    .collectFile(name: 'chr22.bed', newLine: true)
                    .map { file -> [ [ id:'chr22' ], file ] }
                input[3] = true
                input[4] = 3
                """
            }
        }

        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    path(workflow.out.bed.get(0).get(1)).readLines(),
                    workflow.out.bed_intervals.findAll { it[0] instanceof String }.collect { path(it[0]).readLines() },
                ).match() }
            )
        }

    }

    test("[], bed, false, false, 3") {

        when {
            workflow {
                """
                input[0] = [[],[]]
                input[1] = Channel.of('chr22\t0\t1000')
                    .collectFile(name: 'chr22.bed', newLine: true)
                    .map { file -> [ [ id:'chr22' ], file ] }
                input[2] = false
                input[3] = false
                input[4] = 3
                """
            }
        }

        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    path(workflow.out.bed.get(0).get(1)).readLines(),
                    workflow.out.bed_intervals.findAll { it[0] instanceof String }.collect { path(it[0]).readLines() },
                ).match() }
            )
        }

    }

    test("[], bed, false, true, 3") {

        when {
            workflow {
                """
                input[0] = [[],[]]
                input[1] = Channel.of('''chr22\t0\t1000\nchr22\t500\t1500\nchr22\t2000\t3000''')
                    .collectFile(name: 'chr22.bed', newLine: true)
                    .map { file -> [ [ id:'chr22' ], file ] }
                input[2] = false
                input[3] = true
                input[4] = 3
                """
            }
        }

        then {
            assertAll (
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    path(workflow.out.bed.get(0).get(1)).readLines(),
                    workflow.out.bed_intervals.findAll { it[0] instanceof String }.collect { path(it[0]).readLines() },
                ).match() }
            )
        }

    }

    test("[], bed, false, true, 0 - should fail") {

        when {
            workflow {
                """
                input[0] = [[],[]]
                input[1] = Channel.of('''chr22\t0\t1000\nchr22\t500\t1500\nchr22\t2000\t3000''')
                    .collectFile(name: 'chr22.bed', newLine: true)
                    .map { file -> [ [ id:'chr22' ], file ] }
                input[2] = false
                input[3] = true
                input[4] = 0
                """
            }
        }

        then {
            assertAll (
                { assert workflow.failed },
            )
        }

    }

}
