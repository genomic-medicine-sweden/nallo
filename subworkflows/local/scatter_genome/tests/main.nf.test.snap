{
    "fai, [], bed, true, 3": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "genome.fasta.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "1": [
                    [
                        "1.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7",
                        1
                    ]
                ],
                "2": [
                    "versions.yml:md5,8174fa0cc1f95ed4b353a948a2cbd4e9",
                    "versions.yml:md5,beb3374eccd225096e2f7c8ecd3a28be",
                    "versions.yml:md5,d69b1c1fc4e665c357d541ea3b0a0240",
                    "versions.yml:md5,e70d387c0dd30530a3e7fe2846ddffe1"
                ],
                "bed": [
                    [
                        {
                            "id": "test"
                        },
                        "genome.fasta.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "bed_intervals": [
                    [
                        "1.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7",
                        1
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8174fa0cc1f95ed4b353a948a2cbd4e9",
                    "versions.yml:md5,beb3374eccd225096e2f7c8ecd3a28be",
                    "versions.yml:md5,d69b1c1fc4e665c357d541ea3b0a0240",
                    "versions.yml:md5,e70d387c0dd30530a3e7fe2846ddffe1"
                ]
            },
            [
                "chr22\t0\t40001"
            ],
            [
                [
                    "chr22\t0\t40001"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-11T13:03:46.414904212"
    },
    "[], bed, false, false, 3": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "chr22"
                        },
                        "chr22.bed:md5,3b0b598acca89a84aa414e4c95abec1f"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "bed": [
                    [
                        {
                            "id": "chr22"
                        },
                        "chr22.bed:md5,3b0b598acca89a84aa414e4c95abec1f"
                    ]
                ],
                "bed_intervals": [
                    
                ],
                "versions": [
                    
                ]
            },
            [
                "chr22\t0\t1000"
            ],
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-11T12:53:03.541050122"
    },
    "[], bed, false, true, 3": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "chr22"
                        },
                        "chr22.bed:md5,88ebd3fd77139b3d33af226231eff0df"
                    ]
                ],
                "1": [
                    [
                        "1.bed:md5,37978db9b095a29ebbd64f65dd8f13b3",
                        2
                    ],
                    [
                        "2.bed:md5,0da4774e61b9c9005122e46f24522604",
                        2
                    ]
                ],
                "2": [
                    "versions.yml:md5,8174fa0cc1f95ed4b353a948a2cbd4e9",
                    "versions.yml:md5,beb3374eccd225096e2f7c8ecd3a28be",
                    "versions.yml:md5,d69b1c1fc4e665c357d541ea3b0a0240"
                ],
                "bed": [
                    [
                        {
                            "id": "chr22"
                        },
                        "chr22.bed:md5,88ebd3fd77139b3d33af226231eff0df"
                    ]
                ],
                "bed_intervals": [
                    [
                        "1.bed:md5,37978db9b095a29ebbd64f65dd8f13b3",
                        2
                    ],
                    [
                        "2.bed:md5,0da4774e61b9c9005122e46f24522604",
                        2
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8174fa0cc1f95ed4b353a948a2cbd4e9",
                    "versions.yml:md5,beb3374eccd225096e2f7c8ecd3a28be",
                    "versions.yml:md5,d69b1c1fc4e665c357d541ea3b0a0240"
                ]
            },
            [
                "chr22\t0\t1000",
                "chr22\t500\t1500",
                "chr22\t2000\t3000"
            ],
            [
                [
                    "chr22\t0\t1500"
                ],
                [
                    "chr22\t2000\t3000"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-11T13:06:34.382235567"
    },
    "fai, [], true, true, 3": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "genome.fasta.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "1": [
                    [
                        "1.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7",
                        1
                    ]
                ],
                "2": [
                    "versions.yml:md5,8174fa0cc1f95ed4b353a948a2cbd4e9",
                    "versions.yml:md5,beb3374eccd225096e2f7c8ecd3a28be",
                    "versions.yml:md5,d69b1c1fc4e665c357d541ea3b0a0240",
                    "versions.yml:md5,e70d387c0dd30530a3e7fe2846ddffe1"
                ],
                "bed": [
                    [
                        {
                            "id": "test"
                        },
                        "genome.fasta.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "bed_intervals": [
                    [
                        "1.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7",
                        1
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8174fa0cc1f95ed4b353a948a2cbd4e9",
                    "versions.yml:md5,beb3374eccd225096e2f7c8ecd3a28be",
                    "versions.yml:md5,d69b1c1fc4e665c357d541ea3b0a0240",
                    "versions.yml:md5,e70d387c0dd30530a3e7fe2846ddffe1"
                ]
            },
            [
                "chr22\t0\t40001"
            ],
            [
                [
                    "chr22\t0\t40001"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-11T13:03:35.848426979"
    },
    "fai, [], true, false, []": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "genome.fasta.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,e70d387c0dd30530a3e7fe2846ddffe1"
                ],
                "bed": [
                    [
                        {
                            "id": "test"
                        },
                        "genome.fasta.bed:md5,87a15eb9c2ff20ccd5cd8735a28708f7"
                    ]
                ],
                "bed_intervals": [
                    
                ],
                "versions": [
                    "versions.yml:md5,e70d387c0dd30530a3e7fe2846ddffe1"
                ]
            },
            [
                "chr22\t0\t40001"
            ],
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-11T13:03:25.327078297"
    }
}